google/cloud/pubsub_v1/publisher/flow_controller.py

```diff
@@ -204,8 +201,16 @@ def _distribute_available_bytes(self):
 
             reservation = self._byte_reservations[thread]
             still_needed = reservation.needed - reservation.reserved
-            can_give = min(still_needed, available)
 
+            # Sanity check for any internal inconsistencies.
+            if still_needed < 0:
```

Comment:
I usually assert in these buggy cases. Users notice asserts and we fix them. Otherwise, bugs might lurk for a while if no one checks the logs. It's up to you though.

Reply:
Using assert could be an option, but after some thought I would still go with a warning message, as this is the approach already used in leaser, for instance, i.e. log and correct the issue.

Apart from consistency, a warning is probably better suited here, since we have a reasonable chance of recovering from an unexpected state, as opposed to `assert` that would just break immediately (or is just a no-op when Python is run in non-debug mode).

Nevertheless, I will bring up this topic at the next libs meeting to check what other libraries usually do (I know that in BigQuery warnings are used).

Reply:
Sounds reasonable to me. :)